diff -Nru ../orig/qemu-3.1.1.1/accel/kvm/kvm-all.c ./accel/kvm/kvm-all.c
--- ../orig/qemu-3.1.1.1/accel/kvm/kvm-all.c	2019-10-01 15:31:43.000000000 -0700
+++ ./accel/kvm/kvm-all.c	2020-12-11 15:51:03.100843900 -0800
@@ -819,6 +819,41 @@
     }
 }

+void kvm_update_guest_pa_range(uint64_t start_pa, uint64_t size, void *host_va, int readonly, int add)
+{
+    KVMState *s = kvm_state;
+    KVMMemoryListener *kml = &s->memory_listener;
+
+    MemoryRegion mr;
+    MemoryRegionSection section;
+    RAMBlock ram_block;
+
+    memset(&ram_block, 0, sizeof(RAMBlock));
+    ram_block.mr = &mr;
+    ram_block.used_length = HOST_PAGE_ALIGN(size);
+    ram_block.max_length = HOST_PAGE_ALIGN(size);
+    ram_block.fd = -1;
+    ram_block.page_size = getpagesize();
+    ram_block.host = host_va;
+    ram_block.flags |= RAM_PREALLOC;
+
+    memory_region_init(&mr, NULL, NULL, HOST_PAGE_ALIGN(size));
+    mr.ram = true;
+    mr.ram_block = &ram_block;
+    mr.readonly = readonly;
+    mr.nonvolatile = false;
+
+    section.mr = &mr;
+    section.fv = 0;
+    section.offset_within_region = 0;
+    section.size = mr.size;
+    section.offset_within_address_space = start_pa;
+    section.readonly = mr.readonly;
+    section.nonvolatile = mr.nonvolatile;
+
+    kvm_set_phys_mem(kml, &section, add);
+}
+
 static void kvm_region_add(MemoryListener *listener,
                            MemoryRegionSection *section)
 {
diff -Nru ../orig/qemu-3.1.1.1/configure ./configure
--- ../orig/qemu-3.1.1.1/configure	2019-10-01 15:31:44.000000000 -0700
+++ ./configure	2021-07-26 02:31:43.196438200 -0700
@@ -782,6 +782,8 @@
   else
     audio_drv_list=""
   fi
+# For MESA/Glide pass-through ##
+  LIBS="-lgdi32 $LIBS"
   supported_os="yes"
 ;;
 GNU/kFreeBSD)
@@ -881,6 +883,8 @@
   vhost_scsi="yes"
   vhost_vsock="yes"
   QEMU_INCLUDES="-I\$(SRC_PATH)/linux-headers -I$PWD/linux-headers $QEMU_INCLUDES"
+# For MESA/Glide pass-through ##
+  LIBS="-ldl -lXxf86vm -lGL $LIBS"
   supported_os="yes"
   libudev="yes"
 ;;
diff -Nru ../orig/qemu-3.1.1.1/hw/Makefile.objs ./hw/Makefile.objs
--- ../orig/qemu-3.1.1.1/hw/Makefile.objs	2019-10-01 15:31:44.000000000 -0700
+++ ./hw/Makefile.objs	2020-06-01 18:34:16.581486000 -0700
@@ -7,6 +7,7 @@
 devices-dirs-$(CONFIG_SOFTMMU) += char/
 devices-dirs-$(CONFIG_SOFTMMU) += cpu/
 devices-dirs-$(CONFIG_SOFTMMU) += display/
+devices-dirs-$(CONFIG_SOFTMMU) += 3dfx/
 devices-dirs-$(CONFIG_SOFTMMU) += dma/
 devices-dirs-$(CONFIG_SOFTMMU) += gpio/
 devices-dirs-$(CONFIG_HYPERV) += hyperv/
@@ -17,6 +18,7 @@
 devices-dirs-$(CONFIG_IPACK) += ipack/
 devices-dirs-$(CONFIG_IPMI) += ipmi/
 devices-dirs-$(CONFIG_SOFTMMU) += isa/
+devices-dirs-$(CONFIG_SOFTMMU) += mesa/
 devices-dirs-$(CONFIG_SOFTMMU) += misc/
 devices-dirs-$(CONFIG_SOFTMMU) += net/
 devices-dirs-$(CONFIG_SOFTMMU) += rdma/
diff -Nru ../orig/qemu-3.1.1.1/hw/i386/pc.c ./hw/i386/pc.c
--- ../orig/qemu-3.1.1.1/hw/i386/pc.c	2019-10-01 15:31:44.000000000 -0700
+++ ./hw/i386/pc.c	2020-08-14 01:47:55.734263100 -0700
@@ -1636,6 +1636,28 @@
 
     /* Super I/O */
     pc_superio_init(isa_bus, create_fdctrl, no_vmport);
+    /* Glide pass-through */
+    glidept_mm_init();
+    /* MESA pass-through */
+    mesapt_mm_init();
+}
+
+void glidept_mm_init(void)
+{
+    DeviceState *glidept_dev = NULL;
+
+    glidept_dev = qdev_new(TYPE_GLIDEPT);
+    sysbus_realize(SYS_BUS_DEVICE(glidept_dev), &error_fatal);
+    sysbus_mmio_map(SYS_BUS_DEVICE(glidept_dev), 0, GLIDEPT_MM_BASE);
+}
+
+void mesapt_mm_init(void)
+{
+    DeviceState *mesapt_dev = NULL;
+
+    mesapt_dev = qdev_new(TYPE_MESAPT);
+    sysbus_realize(SYS_BUS_DEVICE(mesapt_dev), &error_fatal);
+    sysbus_mmio_map(SYS_BUS_DEVICE(mesapt_dev), 0, MESAPT_MM_BASE);
 }
 
 void pc_nic_init(PCMachineClass *pcmc, ISABus *isa_bus, PCIBus *pci_bus)
diff -Nru ../orig/qemu-3.1.1.1/include/hw/i386/pc.h ./include/hw/i386/pc.h
--- ../orig/qemu-3.1.1.1/include/hw/i386/pc.h	2019-10-01 15:31:45.000000000 -0700
+++ ./include/hw/i386/pc.h	2020-08-14 01:45:02.237365100 -0700
@@ -165,6 +165,22 @@
 
 void gsi_handler(void *opaque, int n, int level);
 
+#if (((QEMU_VERSION_MAJOR << 8) | \
+      (QEMU_VERSION_MINOR << 4) | \
+       QEMU_VERSION_MICRO) < 0x510)
+#define qdev_new(x)         qdev_create(NULL,x)
+#define sysbus_realize(x,y) qdev_init_nofail((DeviceState *)x)
+#endif
+/* GLIDE pass-through */
+#define TYPE_GLIDEPT "glidept"
+#define GLIDEPT_MM_BASE 0xfbdff000
+void glidept_mm_init(void);
+
+/* MESA pass-through */
+#define TYPE_MESAPT "mesapt"
+#define MESAPT_MM_BASE 0xefffe000
+void mesapt_mm_init(void);
+
 /* vmport.c */
 #define TYPE_VMPORT "vmport"
 typedef uint32_t (VMPortReadFunc)(void *opaque, uint32_t address);
diff -Nru ../orig/qemu-3.1.1.1/include/sysemu/kvm.h ./include/sysemu/kvm.h
--- ../orig/qemu-3.1.1.1/include/sysemu/kvm.h	2019-10-01 15:31:45.000000000 -0700
+++ ./include/sysemu/kvm.h	2020-12-11 15:48:07.315383300 -0800
@@ -475,6 +475,8 @@

 #endif /* NEED_CPU_H */

+void kvm_update_guest_pa_range(uint64_t start_pa, uint64_t size, void *host_va, int readonly, int add);
+
 void kvm_cpu_synchronize_state(CPUState *cpu);
 void kvm_cpu_synchronize_post_reset(CPUState *cpu);
 void kvm_cpu_synchronize_post_init(CPUState *cpu);
diff -Nru ../orig/qemu-3.1.1.1/include/sysemu/whpx.h ./include/sysemu/whpx.h
--- ../orig/qemu-3.1.1.1/include/sysemu/whpx.h	2019-10-01 15:31:45.000000000 -0700
+++ ./include/sysemu/whpx.h	2021-07-08 22:14:30.110464300 -0700
@@ -21,6 +21,7 @@
 void whpx_destroy_vcpu(CPUState *cpu);
 void whpx_vcpu_kick(CPUState *cpu);

+void whpx_update_guest_pa_range(uint64_t start_pa, uint64_t size, void *host_va, int readonly, int add);

 void whpx_cpu_synchronize_state(CPUState *cpu);
 void whpx_cpu_synchronize_post_reset(CPUState *cpu);
diff -Nru ../orig/qemu-3.1.1.1/include/ui/console.h ./include/ui/console.h
--- ../orig/qemu-3.1.1.1/include/ui/console.h	2019-10-01 15:31:45.000000000 -0700
+++ ./include/ui/console.h	2021-07-26 01:51:38.502766300 -0700
@@ -464,4 +464,21 @@
 /* input.c */
 int index_from_key(const char *key, size_t key_length);
 
+/* glidewnd.c */
+void *glide_prepare_window(int, int);
+void glide_release_window(void);
+int glide_window_stat(int);
+int glide_gui_fullscreen(int *, int *);
+int is_glide_enabled(void);
+void glide_enabled_set(void);
+void glide_enabled_reset(void);
+
+/* mglcntx.c */
+void *mesa_prepare_window(void);
+void mesa_release_window(void);
+int mesa_gui_fullscreen(int *, int *);
+int is_mesa_enabled(void);
+void mesa_enabled_set(void);
+void mesa_enabled_reset(void);
+
 #endif
diff -Nru ../orig/qemu-3.1.1.1/target/i386/whpx-all.c ./target/i386/whpx-all.c
--- ../orig/qemu-3.1.1.1/target/i386/whpx-all.c	2019-10-01 15:31:46.000000000 -0700
+++ ./target/i386/whpx-all.c	2020-12-11 15:50:20.905823800 -0800
@@ -10,6 +10,7 @@

 #include "qemu/osdep.h"
 #include "cpu.h"
+#include "exec/ram_addr.h"
 #include "exec/address-spaces.h"
 #include "exec/ioport.h"
 #include "qemu-common.h"
@@ -1300,6 +1301,38 @@
                         memory_region_is_rom(mr), mr->name);
 }

+void whpx_update_guest_pa_range(uint64_t start_pa, uint64_t size, void *host_va, int readonly, int add)
+{
+    MemoryRegion mr;
+    MemoryRegionSection section;
+    RAMBlock ram_block;
+
+    memset(&ram_block, 0, sizeof(RAMBlock));
+    ram_block.mr = &mr;
+    ram_block.used_length = HOST_PAGE_ALIGN(size);
+    ram_block.max_length = HOST_PAGE_ALIGN(size);
+    ram_block.fd = -1;
+    ram_block.page_size = getpagesize();
+    ram_block.host = host_va;
+    ram_block.flags |= RAM_PREALLOC;
+
+    memory_region_init(&mr, NULL, NULL, HOST_PAGE_ALIGN(size));
+    mr.ram = true;
+    mr.ram_block = &ram_block;
+    mr.readonly = readonly;
+    mr.nonvolatile = false;
+
+    section.mr = &mr;
+    section.fv = 0;
+    section.offset_within_region = 0;
+    section.size = mr.size;
+    section.offset_within_address_space = start_pa;
+    section.readonly = mr.readonly;
+    section.nonvolatile = mr.nonvolatile;
+
+    whpx_process_section(&section, add);
+}
+
 static void whpx_region_add(MemoryListener *listener,
                            MemoryRegionSection *section)
 {
diff -Nru ../orig/qemu-3.1.1.1/ui/console.c ./ui/console.c
--- ../orig/qemu-3.1.1.1/ui/console.c	2019-10-01 15:31:46.000000000 -0700
+++ ./ui/console.c	2021-08-02 15:06:30.972250800 -0700
@@ -258,12 +258,24 @@
     ds->have_text = have_text;
 }
 
+static int glide_enabled = 0;
+int is_glide_enabled(void) { return glide_enabled; }
+void glide_enabled_set(void) { glide_enabled = 1; }
+void glide_enabled_reset(void) { glide_enabled = 0; }
+
+static int mesa_enabled = 0;
+int is_mesa_enabled(void) { return mesa_enabled; }
+void mesa_enabled_set(void) { mesa_enabled = 1; }
+void mesa_enabled_reset(void) { mesa_enabled = 0; }
+
 void graphic_hw_update(QemuConsole *con)
 {
     if (!con) {
         con = active_console;
     }
     if (con && con->hw_ops->gfx_update) {
+        if (is_glide_enabled() || is_mesa_enabled()) { }
+        else
         con->hw_ops->gfx_update(con->hw);
     }
 }
diff -Nru ../orig/qemu-3.1.1.1/ui/sdl.c ./ui/sdl.c
--- ../orig/qemu-3.1.1.1/ui/sdl.c	2019-10-01 15:31:46.000000000 -0700
+++ ./ui/sdl.c	2021-08-02 15:14:16.596986400 -0700
@@ -541,6 +541,8 @@
         keycode = sdl_keyevent_to_keycode(&ev->key);
         switch (keycode) {
         case 0x21: /* 'f' key on US keyboard */
+            if (is_glide_enabled() || is_mesa_enabled())
+                break;
             toggle_full_screen();
             gui_keysym = 1;
             break;
@@ -822,6 +824,10 @@
             break;
         case SDL_VIDEORESIZE:
             sdl_scale(ev->resize.w, ev->resize.h);
+            if ((ev->resize.w == gui_saved_width) && (ev->resize.h == gui_saved_height)) {
+                gui_saved_width = 0;
+                gui_saved_height = 0;
+            }
             graphic_hw_invalidate(NULL);
             graphic_hw_update(NULL);
             break;
@@ -892,6 +898,112 @@
     SDL_QuitSubSystem(SDL_INIT_VIDEO);
 }
 
+static void sdl_display_valid(const char *feat)
+{
+    if (!guest_screen) {
+        error_report("%s: invalid sdl display. Use -display sdl", feat);
+        exit(1);
+    }
+}
+
+void *glide_prepare_window(int w, int h)
+{
+    void *hwndSDL = NULL;
+    SDL_SysWMinfo wmi;
+
+    sdl_display_valid("glidept");
+
+    SDL_VERSION(&wmi.version);
+    if(SDL_GetWMInfo(&wmi)) {
+#if defined(SDL_VIDEO_DRIVER_WINDIB) || defined(SDL_VIDEO_DRIVER_DDRAW)
+	hwndSDL = (void *)wmi.window;
+#elif defined(SDL_VIDEO_DRIVER_X11)
+	hwndSDL = (void *)wmi.info.x11.window;
+#else
+#error Unsupported window binding
+#endif
+    }
+    gui_saved_width = real_screen->w;
+    gui_saved_height = real_screen->h;
+
+    glide_enabled_set();
+    smp_mb();
+
+    SDL_Event ev1, *ev = &ev1;
+    ev->type = SDL_VIDEORESIZE;
+    ev->resize.w = w;
+    ev->resize.h = h;
+    SDL_PushEvent(ev);
+
+    return hwndSDL;
+}
+
+void glide_release_window(void)
+{
+    glide_enabled_reset();
+    smp_mb();
+    SDL_Event ev1, *ev = &ev1;
+    ev->type = SDL_VIDEORESIZE;
+    ev->resize.w = gui_saved_width;
+    ev->resize.h = gui_saved_height;
+    SDL_PushEvent(ev);
+}
+
+int glide_window_stat(int activate)
+{
+    int scr_w, scr_h, stat;
+    scr_w = real_screen->w;
+    scr_h = real_screen->h;
+    if (activate)
+	stat = (((scr_h & 0x7FFFU) << 0x10) | scr_w);
+    else
+	stat = (((gui_saved_height & 0x7FFFU) << 0x10) | gui_saved_width)? 1:0;
+    return stat;
+}
+
+int glide_gui_fullscreen(int *width, int *height)
+{
+    sdl_display_valid("glidept");
+    if (width)
+        *width = real_screen->w;
+    if (height)
+        *height = real_screen->h;
+    return gui_fullscreen;
+}
+
+void *mesa_prepare_window(void)
+{
+    void *hwndSDL = NULL;
+    SDL_SysWMinfo wmi;
+
+    sdl_display_valid("mesapt");
+
+    SDL_VERSION(&wmi.version);
+    if(SDL_GetWMInfo(&wmi)) {
+#if defined(SDL_VIDEO_DRIVER_WINDIB) || defined(SDL_VIDEO_DRIVER_DDRAW)
+	hwndSDL = (void *)wmi.window;
+#elif defined(SDL_VIDEO_DRIVER_X11)
+	hwndSDL = (void *)wmi.info.x11.window;
+#else
+#error Unsupported window binding
+#endif
+    }
+
+    return hwndSDL;
+}
+
+void mesa_release_window(void) { }
+
+int mesa_gui_fullscreen(int *width, int *height)
+{
+    sdl_display_valid("mesapt");
+    if (width)
+        *width = real_screen->w;
+    if (height)
+        *height = real_screen->h;
+    return gui_fullscreen;
+}
+
 static const DisplayChangeListenerOps dcl_ops = {
     .dpy_name             = "sdl",
     .dpy_gfx_update       = sdl_update,
diff -Nru ../orig/qemu-3.1.1.1/ui/sdl2.c ./ui/sdl2.c
--- ../orig/qemu-3.1.1.1/ui/sdl2.c	2019-10-01 15:31:46.000000000 -0700
+++ ./ui/sdl2.c	2021-08-02 15:07:26.817890200 -0700
@@ -36,6 +36,8 @@
 static SDL_Surface *guest_sprite_surface;
 static int gui_grab; /* if true, all keyboard/mouse events are grabbed */
 
+static int gui_saved_width;
+static int gui_saved_height;
 static int gui_saved_grab;
 static int gui_fullscreen;
 static int gui_keysym;
@@ -359,6 +361,8 @@
             }
             break;
         case SDL_SCANCODE_F:
+            if (is_glide_enabled() || is_mesa_enabled())
+                break;
             toggle_full_screen(scon);
             gui_keysym = 1;
             break;
@@ -535,6 +539,14 @@
         sdl2_redraw(scon);
         break;
     case SDL_WINDOWEVENT_EXPOSED:
+	if (ev->window.data1 && ev->window.data2) {
+	    if ((ev->window.data1 == gui_saved_width) &&
+		(ev->window.data2 == gui_saved_height)) {
+		gui_saved_width = 0;
+		gui_saved_height = 0;
+	    }
+	    SDL_SetWindowSize(scon->real_window, ev->window.data1, ev->window.data2);
+	}
         sdl2_redraw(scon);
         break;
     case SDL_WINDOWEVENT_FOCUS_GAINED:
@@ -719,6 +731,117 @@
     SDL_QuitSubSystem(SDL_INIT_VIDEO);
 }
 
+static void sdl_display_valid(const char *feat)
+{
+    if (!sdl2_console) {
+        error_report("%s: invalid sdl display. Use -display sdl", feat);
+        exit(1);
+    }
+}
+
+void *glide_prepare_window(int w, int h)
+{
+    void *hwndSDL = NULL;
+    SDL_SysWMinfo wmi;
+    SDL_Event ev1, *ev = &ev1;
+    int scr_w, scr_h;
+
+    sdl_display_valid("glidept");
+
+    SDL_VERSION(&wmi.version);
+    if (SDL_GetWindowWMInfo(sdl2_console[0].real_window, &wmi)) {
+#if defined(SDL_VIDEO_DRIVER_WINDOWS)	
+	hwndSDL = (void *)wmi.info.win.window;
+#elif defined(SDL_VIDEO_DRIVER_X11)
+	hwndSDL = (void *)wmi.info.x11.window;
+#else
+#error Unsupported window binding
+#endif
+    }
+    SDL_GetWindowSize(sdl2_console[0].real_window, &scr_w, &scr_h);
+    gui_saved_width = scr_w;
+    gui_saved_height = scr_h;
+
+    glide_enabled_set();
+    smp_mb();
+
+    ev->type = SDL_WINDOWEVENT;
+    ev->window.windowID = SDL_GetWindowID(sdl2_console[0].real_window);
+    ev->window.event = SDL_WINDOWEVENT_EXPOSED;
+    ev->window.data1 = w;
+    ev->window.data2 = h;
+    SDL_PushEvent(ev);
+
+    return hwndSDL;
+}
+
+void glide_release_window(void)
+{
+    glide_enabled_reset();
+    smp_mb();
+    SDL_Event ev1, *ev = &ev1;
+    ev->type = SDL_WINDOWEVENT;
+    ev->window.windowID = SDL_GetWindowID(sdl2_console[0].real_window);
+    ev->window.event = SDL_WINDOWEVENT_EXPOSED;
+    ev->window.data1 = gui_saved_width;
+    ev->window.data2 = gui_saved_height;
+    SDL_PushEvent(ev);
+}
+
+int glide_window_stat(int activate)
+{
+    int scr_w, scr_h, stat;
+    SDL_GetWindowSize(sdl2_console[0].real_window, &scr_w, &scr_h);
+    if (activate)
+	stat = (((scr_h & 0x7FFFU) << 0x10) | scr_w);
+    else
+	stat = (((gui_saved_height & 0x7FFFU) << 0x10) | gui_saved_width)? 1:0;
+    return stat;
+}
+
+int glide_gui_fullscreen(int *width, int *height)
+{
+    sdl_display_valid("glidept");
+    if (width)
+        *width = surface_width(sdl2_console[0].surface);
+    if (height)
+        *height = surface_height(sdl2_console[0].surface);
+    return gui_fullscreen;
+}
+
+void *mesa_prepare_window(void)
+{
+    void *hwndSDL = NULL;
+    SDL_SysWMinfo wmi;
+
+    sdl_display_valid("mesapt");
+
+    SDL_VERSION(&wmi.version);
+    if (SDL_GetWindowWMInfo(sdl2_console[0].real_window, &wmi)) {
+#if defined(SDL_VIDEO_DRIVER_WINDOWS)	
+	hwndSDL = (void *)wmi.info.win.window;
+#elif defined(SDL_VIDEO_DRIVER_X11)
+	hwndSDL = (void *)wmi.info.x11.window;
+#else
+#error Unsupported window binding
+#endif
+    }
+
+    return hwndSDL;
+}
+
+void mesa_release_window(void) { }
+
+int mesa_gui_fullscreen(int *width, int *height)
+{
+    sdl_display_valid("mesapt");
+    if (width)
+        *width = surface_width(sdl2_console[0].surface);
+    if (height)
+        *height = surface_height(sdl2_console[0].surface);
+    return gui_fullscreen;
+}
+
 static const DisplayChangeListenerOps dcl_2d_ops = {
     .dpy_name             = "sdl2-2d",
     .dpy_gfx_update       = sdl2_2d_update,
diff -Nru ../orig/qemu-3.1.1.1/vl.c ./vl.c
--- ../orig/qemu-3.1.1.1/vl.c	2019-10-01 15:31:47.000000000 -0700
+++ ./vl.c	2021-07-29 17:27:27.411089200 -0700
@@ -1896,6 +1896,13 @@
     }
 }

+static void feature(void)
+{
+    const char rev_[ALIGNED(1)]
+        ;
+    printf("  featuring qemu-3dfx@%s"__TIME__" "__DATE__" build\n", rev_);
+}
+
 static void version(void)
 {
     printf("QEMU emulator version " QEMU_FULL_VERSION "\n"
@@ -3292,6 +3308,7 @@
                 break;
             case QEMU_OPTION_version:
                 version();
+                feature();
                 exit(0);
                 break;
             case QEMU_OPTION_m:
