diff -Nru ../orig/qemu-7.2.0/ui/sdl2.c ./ui/sdl2.c
--- ../orig/qemu-7.2.0/ui/sdl2.c
+++ ./ui/sdl2.c
@@ -96,6 +96,10 @@
     }
 #ifdef CONFIG_OPENGL
     if (scon->opengl) {
+#ifdef CONFIG_DARWIN
+        /* Apple OpenGL quirk */
+        SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
+#endif
         flags |= SDL_WINDOW_OPENGL;
     }
 #endif
diff -Nru ../orig/qemu-6.1.0/ui/shader/texture-blit-flip.vert ./ui/shader/texture-blit-flip.vert
--- ../orig/qemu-6.1.0/ui/shader/texture-blit-flip.vert
+++ ./ui/shader/texture-blit-flip.vert
@@ -1,5 +1,3 @@
-#version 300 es
-
 in vec2  in_position;
 out vec2 ex_tex_coord;
 
diff -Nru ../orig/qemu-6.1.0/ui/shader/texture-blit.frag ./ui/shader/texture-blit.frag
--- ../orig/qemu-6.1.0/ui/shader/texture-blit.frag
+++ ./ui/shader/texture-blit.frag
@@ -1,5 +1,3 @@
-#version 300 es
-
 uniform sampler2D image;
 in  mediump vec2 ex_tex_coord;
 out mediump vec4 out_frag_color;
diff -Nru ../orig/qemu-6.1.0/ui/shader/texture-blit.vert ./ui/shader/texture-blit.vert
--- ../orig/qemu-6.1.0/ui/shader/texture-blit.vert
+++ ./ui/shader/texture-blit.vert
@@ -1,5 +1,3 @@
-#version 300 es
-
 in vec2  in_position;
 out vec2 ex_tex_coord;
 
diff -Nru ../orig/qemu-6.1.0/ui/shader.c ./ui/shader.c
--- ../orig/qemu-6.1.0/ui/shader.c
+++ ./ui/shader.c
@@ -152,11 +152,22 @@
 QemuGLShader *qemu_gl_init_shader(void)
 {
     QemuGLShader *gls = g_new0(QemuGLShader, 1);
+    const GLchar *header = epoxy_is_desktop_gl() ? "#version 140\n" : "#version 300 es\n";
+    char *frag_src = g_new0(GLchar, 1 + strlen(header) + sizeof(texture_blit_frag_src));
+    char *vert_src = g_new0(GLchar, 1 + strlen(header) +
+            MAX(sizeof(texture_blit_vert_src), sizeof(texture_blit_flip_vert_src)));
+    char *vert_src_body = g_stpcpy(vert_src, header);
+    char *frag_src_body = g_stpcpy(frag_src, header);
 
+    strcpy(vert_src_body, texture_blit_vert_src);
+    strcpy(frag_src_body, texture_blit_frag_src);
     gls->texture_blit_prog = qemu_gl_create_compile_link_program
-        (texture_blit_vert_src, texture_blit_frag_src);
+        (vert_src, frag_src);
+    strcpy(vert_src_body, texture_blit_flip_vert_src);
     gls->texture_blit_flip_prog = qemu_gl_create_compile_link_program
-        (texture_blit_flip_vert_src, texture_blit_frag_src);
+        (vert_src, frag_src);
+    g_free(vert_src);
+    g_free(frag_src);
     if (!gls->texture_blit_prog || !gls->texture_blit_flip_prog) {
         exit(1);
     }
